lib/src/query.c

```diff
@@ -3541,8 +3542,10 @@ static inline bool ts_query_cursor__advance(
         point_gte(ts_node_start_point(parent_node), self->end_point)
       );
       bool node_precedes_range = parent_precedes_range || (
-        ts_node_end_byte(node) <= self->start_byte ||
-        point_lte(ts_node_end_point(node), self->start_point)
+        !is_empty && (
+          ts_node_end_byte(node) <= self->start_byte ||
+          point_lte(ts_node_end_point(node), self->start_point)
+        )
```

Comment:
I think this would falsely *include* the node if it is empty, but disjoint from the range. For example, say we had an empty node at 0, but were looking for matches in the range `5 - 10`. I think that instead of special-casing byte offset 0, we may want to have a special-case where the node is empty and touches the range boundary.

What about something like this?

```c
uint32_t start_byte = ts_node_start_byte(node);
uint32_t end_byte = ts_node_end_byte(node);
TSPoint start_point = ts_node_start_point(node);
TSPoint end_point = ts_node_end_point(node);
bool is_empty = start_byte == end_byte;

// ...

bool node_precedes_range =
  parent_precedes_range ||
  end_byte < self->start_byte ||
  point_lt(end_point, self->start_point) ||
  (!is_empty && end_byte == self->start_byte) ||
  (!is_empty && point_eq(end_point, self->start_point));
```

Reply:
This makes total sense, and is definitely a lot clearer too